### PR TITLE
add support for validation generators in fit_generator for Sequential and Graph

### PR DESCRIPTION
**Pull request by qdbp**
This is a minimal set of additions to support the use of generators for validation data in `fit_generator ` in both Graph and Sequential. ~~This does not go as far as to implement an `evaluate_generator` method (which would allow, presumably, to use validation generators in other fit methods as well).~~
In addition to the above functionality, an `evaluate_generator` method was also added.

Basic tests were added (and passed) in `test_models.py`

### CODE REVIEW

File: keras/models.py [hunk 1]
```diff
@@ -856,7 +856,9 @@ def load_weights(self, filepath):
 
     def fit_generator(self, generator, samples_per_epoch, nb_epoch,
                       verbose=1, show_accuracy=False, callbacks=[],
-                      validation_data=None, class_weight=None, nb_worker=1):
+                      val_generator=None,
```
Comment (fchollet) [inline keras/models.py]: In terms of API, I think `validation_data` should be overloaded to support passing a generator. No need for a need keyword argument.

File: keras/models.py [hunk 2]
```diff
@@ -1357,8 +1406,10 @@ def fit_generator(self, generator, samples_per_epoch, nb_epoch,
             callbacks: list of callbacks to be called during training.
             validation_data: dictionary mapping input names and outputs names
                 to appropriate numpy arrays to be used as
-                held-out validation data.
-                All arrays should contain the same number of samples.
+                held-out validation data, or a generator yielding such
+                dictionaries. All arrays should contain the same number
+                of samples. If a generator, a single yield will be used
```
Comment (jerheff) [inline keras/models.py]: How do you feel about adding a parameter for the number of seen examples to use in evaluating the validation loss?   This would mirror the functionality of the samples_per_epoch parameter.

Comment (neggert) [inline keras/models.py]: +1. Use case would be validating on more data than fits in GPU memory. So the generator has to return small-ish minibatches, but I want to validate on more than one minibatch.

Comment (qdbp) [inline keras/models.py]: That makes sense.

However, with such an addition I feel it will be more principled to go ahead and implement `evaluate_generator` and `predict_generator`. I think I will pursue this approach.

Though, perhaps more principled still would be universal generator transparency, i.e. use of generators in basic `fit`, `predict` and `evaluate`, freely mixing data and validation arrays and generators.

File: keras/models.py [hunk 3]
```diff
@@ -883,10 +884,13 @@ def fit_generator(self, generator, samples_per_epoch, nb_epoch,
...
-            validation_data: tuple of 2 or 3 numpy arrays. If 2 elements,
-                they are assumed to be (input_data, target_data);
+            validation_data: tuple of 2 or 3 numpy arrays, or a generator.
+                If 2 elements, they are assumed to be (input_data, target_data);
                 if 3 elements, they are assumed to be
-                (input_data, target_data, sample weights).
+                (input_data, target_data, sample weights). If generator,
+                it is assumed to yield tuples of 2 or 3 elements as above.
+                A single batch yielded by this generator will be used each
+                epoch.
```
Comment (fchollet) [inline keras/models.py]: This seems like very strange behavior. The use case of using a generator for validation data is if your validation dataset is too large to fit in memory. In that case you need to break it up into many batches. Running eval on only one batch defeats the purpose. Better would be to run it on arbitrarily many batches until we reach some fixed number of samples, defined in the constructor (e.g. `validation_samples_per_epoch`).

File: keras/models.py [hunk 4]
```diff
@@ -856,7 +856,8 @@ def load_weights(self, filepath):
 
     def fit_generator(self, generator, samples_per_epoch, nb_epoch,
                       verbose=1, show_accuracy=False, callbacks=[],
-                      validation_data=None, class_weight=None, nb_worker=1):
+                      validation_data=None, class_weight=None,
+                      nb_worker=1, nb_val_worker=1):
```
Comment (fchollet) [inline keras/models.py]: `nb_val_worker` should be described in the docstring.

File: keras/models.py (hunk 2 above)
Comment (fchollet) [inline keras/models.py]: Having `evaluate_generator` seems reasonable. It would not be easily feasible to include generator support in `fit` though (since the training data is split into two arguments).
